### ISSUE TITLE
Translate semantic graph FlatBuffers into Go semantic graph

### ISSUE THREAD

**Issue by wolffcm** [influxdata/flux#2091]
This issue is analogous to #2005, and can largely follow the example laid out there when it is complete.

DOD:
Tests exist that show each kind of semantic graph node in the FlatBuffers representation maps to a proper "natural" semantic graph node in Go.

*NOTE:* This issue does not include deserializing the Flux type, which will be a separate issue.

**NOTE:** Deserialization code was merged in #2177, but this issue should stay open until a roundtrip test is merged in a separate PR, which can be done once #2090 is complete.

**Comment by jlapacik** [comment]
@wolffcm I believe will require updating the Go semantic nodes with type information to have parity with the Rust semantic graph.

**Comment by nathanielc** [comment]
@jlapacik Makes sense, we have this issue for that https://github.com/influxdata/flux/issues/1960

Tangentially I was thinking that we would just replace uses of the Go semantic types with their FlatBuffer equivalents in the rest of the code. That way we don't have to deserialize the FlatBuffers on the Go side. Alternatively we could do the translation first and refactor it later but if we have to change the Go semantic types anyways maybe it makes sense to just refactor them together? Thoughts?

**Comment by wolffcm** [comment]
> Tangentially I was thinking that we would just replace uses of the Go semantic types with their FlatBuffer equivalents in the rest of the code.

@nathanielc I like this idea.  There are a lot of usages of the semantic package though, especially in the `interpreter` and `compiler` packages, and also many tests.  It would be an enormous commit and hard to isolate bugs.  I think we would want to break up this work somehow.

**Comment by nathanielc** [comment]
Yeah its a ton of work. 

Ok here is what I am think for sequencing:

1. Write code to convert FlatBuffers to Go semantic types.
2. Write the FlatBuffers equivalent of the Flux types (i.e libflux/src/semantic/types.rs) for both Go and Rust.
3. Get all tests passing using Rust parser + type inference. Using the FlatBuffer Flux type structs anytime we need to reference the types of semantic node (instead of the current Go structs for types)
4. Slowly refactor each other system (interpreter, compiler etc) to use the FlatBuffers for the semantic types.

In short: do not create new Go types for the Flux types, just use Flatbuffers for that initially. And then use the Go semantic types until we have it all working, then piece mail replace each component still using Go semantic types. Thoughts?

**Comment by wolffcm** [comment]
Okay, I think I'm getting what you mean.  I got a little confused by whether "types" refers to the Go types that make the semantic graph (like implementors of `Node`), or Flux types.

Here's how I understand your sequencing:

This issue should just be step one of your sequencing, i.e., translate the semantic graph (but not any of  the Flux types) from FlatBuffers to Go natural structs.  Then we could, if we needed, run type inference on the analyzed semantic graph produced by Rust.

Step two could get its own issue, "Design FlatBuffers type schema".

Step three would be dependent on type inference being complete in Rust.  For that issue we would need to replace any instances of the Go natural Flux types with their FlatBuffers equivalents.

Then we could have one issue each to refactor the other systems.

Do I have this right?

**Comment by nathanielc** [comment]
@wolffcm yes that is correct.

**Comment by nathanielc** [comment]
There are three main phases to consuming the type inference algorithm from Rust:

1. Use Rust parsing and analyze steps, do not use Rust type inference
2. Use the complete Rust pipeline (parse, analyze, type check).
3. Remove Go natural semantic struct

## Phase 1

This phase enable us to use Rust as a parser and analyzer and keep using the natural Go semantic struct.

The required pieces are to be able to translate Rust semantic graph to Flatbuffers and decode those Flatbuffers to Go natural semantic structs, without any monotype information.

This means that we still run type inference on the Go side once we retrieve the semantic Flatbuffer from Rust.

## Phase 2

This phase enables us to use all of the Rust logic, the parser, analyzer and type checker. This will mean that the Go equivalents can be removed.

This requires that we can translate Rust semantic graph to Flatbuffers with monotype information and that the Go side can consume that type information

Phase 2 has two steps:

1. First use Go natural semantic structs with the Flatbuffer monotypes
2. Replace usages of Go natural semantic structs with the Flatbuffer semantic structs with their monotype information.

## Phase 3

Once the code in the Go semantic package is no longer used delete it.

**Comment by jlapacik** [comment]
I think as a first step we should deserialize the semantic flatbuffer representation entirely into the natural Go semantic graph - monotypes, polytypes, and all. In my opinion it will be a much easier refactor if we can keep the Go semantic graph around, get type inference working, and then replace the Go semantic graph with its flatbuffer equivalent all at once. It's also a natural way to sequence the work.

Note this means we will need to define `MonoType` and `PolyType` structs in Go that are analogous to the ones defined in Rust. However I think this is a much faster path to integration than if some parts of the semantic graph reference flatbuffer types and some don't.

**Comment by wolffcm** [comment]
@jlapacik

Note, I've already started work that associates (flatbuffer) types with each expression node, here:
https://github.com/influxdata/flux/pull/2277

I'm not certain I completely understand your comment, but... What if we had Go struct types that are just wrappers around flatbuffers Go structs for Flux types (fbsemantic.PolyType for polytypes, fbsemantic.Arr, fbsemantic.Fun, etc for monotypes).  They could be backed by either a `[]byte` that is memory from Rust or copied into Go-managed memory (with the latter to start).  They aren't really much different from normal Go structs, save that the fields need to be accessed via methods.  Does that fit with what you're proposing?

**Comment by jlapacik** [comment]
@wolffcm I think that fits. Although my personal preference would be that they are not wrappers. I'm not convinced that we should immediately ditch the Go semantic graph. I think that requires some more discussion. So, at least initially, I would prefer that the natural and flatbuffer types aren't mixed.

That's just my personal preference but that doesn't really impact the global refactor I suppose.
